### PR TITLE
ENH: scil scoring unique streamlines

### PR DESCRIPTION
**Pull request by AntoineTheb**
This PR is to add an option (`--unique`) to the scil scoring script to assign streamlines to the first bundle (valid or invalid) they fit in and not to all corresponding bundles (in the case of overlapping ROIs). This brings the behavior of the script closer to the Tractometer and makes the "VC/IC/NC" ratios to be accurate. 

Data to test: https://drive.google.com/file/d/1WpZMrq07Dv1eehVceMWQlRtTfH0BPQEr/view?usp=sharing

Ground truth FiberCup with itself:

```bash 
scil_score_tractogram.py FiberCupGroundTruth_all.tck \ 
~/braindata/databases/fibercup/simulated/MICCAI2015/scoring_data/scil_scoring_config.json \
prob_tracking_test \
--gt_dir ~/braindata/databases/fibercup/simulated/MICCAI2015/scoring_data \
--reference ~/braindata/databases/fibercup/simulated/MICCAI2015/masks/wm.nii.gz \
--compute_ic --unique --no_empty -v
```

Prob tracking on the FiberCup:

```bash 
scil_score_tractogram.py tracking_prob_interface_300_075.trk \ 
~/braindata/databases/fibercup/simulated/MICCAI2015/scoring_data/scil_scoring_config.json \ 
prob_tracking_test \
--gt_dir ~/braindata/databases/fibercup/simulated/MICCAI2015/scoring_data \ 
--reference ~/braindata/databases/fibercup/simulated/MICCAI2015/masks/wm.nii.gz \ 
--compute_ic --unique --no_empty -v
```

Det tracking on the FiberCup:

```bash 
scil_score_tractogram.py tracking_det_interface_300_075.trk \
~/braindata/databases/fibercup/simulated/MICCAI2015/scoring_data/scil_scoring_config.json \
det_tracking_test \
--gt_dir ~/braindata/databases/fibercup/simulated/MICCAI2015/scoring_data \
--reference ~/braindata/databases/fibercup/simulated/MICCAI2015/masks/wm.nii.gz \
--compute_ic --unique --no_empty -v
```

VS/IC/NC ratios should sum to 1.

@EmmaRenauld @arnaudbore

### CODE REVIEW

**Comment by scil-admin** [comment]
Build passed ! Good Job :beers: !

**Comment by scil-admin** [comment]
Build passed ! Good Job :beers: !

**Review by EmmaRenauld — COMMENTED** [review]
Some comments first, but I have not tested yet.

File: scripts/scil_score_tractogram.py [hunk 1]
```diff
@@ -419,17 +422,28 @@ def compute_vb_vs_all_bundles(
...
     bundles_stats = []
+
+    all_ids = np.arange(len(sft))
     for i in range(nb_bundles):
         head_filename = gt_heads[i]
         tail_filename = gt_tails[i]
 
         # Extract true connection
         vs_ids, wpc_ids, bundle_stats = \
             extract_vb_vs(
```
[the remark this answers is not in the document]
Comment (AntoineTheb) [reply scripts/scil_score_tractogram.py:433]: Not sure what you mean by "wrong returns" in the docstring as no return is mentioned ... In fact almost no function in there mention returns (and arguments). I can add them if you want

File: scripts/scil_score_tractogram.py [hunk 2]
```diff
@@ -419,17 +422,28 @@ def compute_vb_vs_all_bundles(
...
                 limits_inv_masks[i], args.dilate_endpoints)
 
+        if args.unique:
+            not_vs = np.setdiff1d(np.arange(len(all_ids)), vs_ids)
+            vs_ids = all_ids[vs_ids]
+
+            not_wpc = np.setdiff1d(np.arange(len(all_ids)), wpc_ids)
+            wpc_ids = all_ids[wpc_ids]
+            not_vs_wps = np.intersect1d(not_vs, not_wpc)
+            all_ids = all_ids[not_vs_wps]
```
[the remark this answers is not in the document]
Comment (AntoineTheb) [reply scripts/scil_score_tractogram.py:445]: Problem is that if a streamline is a wpc for more than 1 bundle and a VC for another (or more), I would still end up with duplicates. I would need to check if a streamline is a duplicate wpc and then check if it is a duplicate vs. It's simpler to just assign it to the first bundle where it fits. I agree that I can skip the loops, then.

File: scripts/scil_score_tractogram.py [hunk 3]
```diff
@@ -574,7 +593,7 @@ def compute_tractometry(all_vs_ids, all_wpc_ids, all_ic_ids, all_nc_ids,
     final_results = {
         "tractogram_filename": str(args.in_tractogram),
         "total_streamlines": total_count,
-        "VB": len([x for x in vs_ids_list if len(x) > 0]),
+        "VB": len(vb_sft_list),
```
[the remark this answers is not in the document]
Comment (AntoineTheb) [reply scripts/scil_score_tractogram.py:596]: Ah sorry, I thought they wouldn't be included if there were 0 streamlines included. I'll revert back the change.

File: scripts/scil_score_tractogram.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (EmmaRenauld) [reply scripts/scil_score_tractogram.py:433]: In extract_vb_vs:

```
Returns
    -------
    tc_sft: StatefulTractogram
        SFT of true connections.
    wpc_sft: StatefulTractogram
        SFT of wrong-path-connections.
    fc_sft: StatefulTractogram
        SFT of false connections (streamlines that are too long).
    nc_streamlines: StatefulTractogram
        SFT of no connections (streamlines that loop)
    sft: StatefulTractogram
        SFT of remaining streamlines.
```

Is wrong, now it only returns 3 lists.

File: scripts/scil_score_tractogram.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (EmmaRenauld) [reply scripts/scil_score_tractogram.py:445]: @mdesco

File: scripts/scil_score_tractogram.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (AntoineTheb) [reply scripts/scil_score_tractogram.py:434]: Ah sorry I misinterpreted which method "that" referred to

**Comment by scil-admin** [comment]
Build passed ! Good Job :beers: !

**Comment by scil-admin** [comment]
Build passed ! Good Job :beers: !

**Comment by EmmaRenauld** [comment]
I ran it on my ismrm data. I picked a submission at random so I don't know if it contained duplicates, but the script runs fine **(with a rebase on the master!!)**.

I tagged Max higher, to see with him what you want to do. But based on my experience, I will however come back stronger with what I said previously. I still think you should first verify that a wpc streamline is not a true connection of another bundle before removing it. Actually, I don't even compute wpc anymore when I use this script, it's often not clear what they represent. Ex below, in blue, the WPC of the cingulum is actually the uncinate fasciculus. Considering that in my config file, the UF comes after the cingulum, I will loose a lot of UF if I use your method! (In this case, I recover 43 streamlines / 429). See option remove_wpc_belonging_to_another_bundle; maybe some of what you need is already coded.

This makes me think, you should add a warning under the option `--unique` with: "Careful, order of bundles in the config file will influence your results".
 

image

**Comment by scil-admin** [comment]
Build Failed :boom: 
https://jenkins.scil.usherbrooke.ca/job/scilpy/job/PR-592/6/

**Comment by scil-admin** [comment]
Build Failed :boom: 
https://jenkins.scil.usherbrooke.ca/job/scilpy/job/PR-592/7/

**Comment by scil-admin** [comment]
Build passed ! Good Job :beers: !

**Comment by AntoineTheb** [comment]
@EmmaRenauld can you send me the files you have used for your test of WPC vs VS ? I think I have a solution

**Comment by EmmaRenauld** [comment]
It should work with 

```
scoring_data=$BrainData/databases/ismrm_2015/derivatives/scoring_data/output/version_2022
ref=$scoring_data/t1.nii.gz
config_file=$scoring_data/scoring_config.json

submissions_dir=$BrainData/databases/ismrm_2015/derivatives/submissions/output/anonymized_submissions_converted_tck
tractogram=$submissions_dir/1_0.tck

out_dir=test
mkdir test

scil_score_tractogram.py -v --no_empty --save_wpc_separately --gt_dir $scoring_data --reference $ref $tractogram $config_file $out_dir
```

With team 1_0, I get 18 duplicates between CC and Cingulum for the valid bundles so you can check if it works well +
for WPC, the cingulum_right_WPC contains parts of the valid bundles : UF right + SLF right + cingulum_left.

File: scripts/scil_score_tractogram.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (arnaudbore) [reply scripts/scil_score_tractogram.py]: I sent a message to @mdesco . I agree with @EmmaRenauld you should not be removing a wps if it's a vs for another bundle

**Review by arnaudbore — CHANGES_REQUESTED** [review]
Need some work and maybe a quick discussion.

File: scripts/scil_score_tractogram.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (mdesco) [reply scripts/scil_score_tractogram.py]: Also agree with @EmmaRenauld and @arnaudbore. Happy to have a short call to make sure we are the same page. @AntoineTheb I let you decide.

File: scripts/scil_score_tractogram.py [hunk 4]
```diff
@@ -446,6 +464,19 @@ def compute_vb_vs_all_bundles(
         logging.info("Bundle {}: nb VS = {}"
                      .format(bundle_names[i], bundle_stats["VS"]))
 
+    if args.remove_wpc_belonging_to_another_bundle and args.unique:
+        print(wpc_ids_list)
```
[the remark this answers is not in the document]
Comment (AntoineTheb) [reply scripts/scil_score_tractogram.py:468]: Woops sorry these were testing prints.

File: scripts/scil_score_tractogram.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (AntoineTheb) [reply scripts/scil_score_tractogram.py]: @EmmaRenauld @arnaudbore @mdesco After speaking with Emmanuelle offline I agree with your points. I have added a potential fix in 10c56df7ad11241295ba25cf9adadd144aa2ab51, still waiting on data to test it.

**Comment by scil-admin** [comment]
Build passed ! Good Job :beers: !

**Comment by AntoineTheb** [comment]
@EmmaRenauld @mdesco @arnaudbore I have added the modifications to assign WPCs to VCs (or actually, to assign to both and then filter out WPCs which are also VCs). 

As we can see in this example, the left cingulum does not end up with UF streamlines:

image

While developing, I temporarily added checks to make sure that there were no duplicate streamlines. These checks have been removed prior to pushing.

**Comment by scil-admin** [comment]
Build Failed :boom: 
https://jenkins.scil.usherbrooke.ca/job/scilpy/job/PR-592/10/

**Comment by scil-admin** [comment]
Build Failed :boom: 
https://jenkins.scil.usherbrooke.ca/job/scilpy/job/PR-592/11/

**Comment by AntoineTheb** [comment]
@EmmaRenauld looking at the comments to update them (don't worry I will), I realize that the updates change the behavior of the script: WPCs are either saved separately, or included as VS. However, it seems that they were previously never considered VS ? Is this something we want ? I feel this should be discussed.

**Comment by scil-admin** [comment]
Build passed ! Good Job :beers: !

**Comment by scil-admin** [comment]
Build passed ! Good Job :beers: !

**Comment by scil-admin** [comment]
Build Failed :boom: 
https://jenkins.scil.usherbrooke.ca/job/scilpy/job/PR-592/15/

**Comment by scil-admin** [comment]
Build passed ! Good Job :beers: !

**Comment by scil-admin** [comment]
Build passed ! Good Job :beers: !

**Comment by EmmaRenauld** [comment]
@AntoineTheb  PR#600 was merged, you should merge with your PR to be sure.

**Comment by scil-admin** [comment]
Build passed ! Good Job :beers: !

**Comment by scil-admin** [comment]
Build passed ! Good Job :beers: !

**Comment by arnaudbore** [comment]
Build Failed :boom: 
https://jenkins.scil.usherbrooke.ca/job/scilpy/job/PR-592/19/

**Comment by AntoineTheb** [comment]
Closing in favor of

**Pull request closed without merging**